## 오류처리

```swift
// 오류표현

// 자판기 동작 오류의 종류를 표현한 VendingMachineError 열거형
enum VendingMachineError: Error {
    case invalidInput
    case insufficientFunds(moneyNeeded: Int)
    case outOfStock
}
```

```swift
// 함수에서 발생한 오류 던지기
// 자판기 동작 도중 발생한 오류 던지기
// 오류 발생의 여지가 있는 메서드는 throws를 사용하여 오류를 내포하는 함수임을 표시
class VendingMachine {
    let itemPrice: Int = 100
    let itemCount: Int = 5
    let deposited: Int = 0

    // 돈 받기 메서드
    func receiverMoney(_ money: Int) throws {
        //  입력한 돈이 0이하면 오류를 던짐
        guard money > 0 else {
            throw VendingMachineError.invalidInput
        }
        self.deposited += money
        print("\(money)원 받음")
    }
    // 물건 팔기 메서드
    func vend(numberOfItems numberOfItemsToVend: Int) throws -> String {
        // 원하는 아이템의 수량이 잘못 입력되었으면 오류를 던짐
        guard numberOfItemsToVend > 0 else {
            throw VendingMachineError.invalidInput
        }
        // 구매하려는 수량보다 미리 넣어둔 돈이 적으면 오류를 던짐
        guard numberOfItemsToVend * itemPrice <= deposited else {
            let moneyNeeded: Int
            moneyNeeded = numberOfItemsToVend * itemPrice - deposited

            throw VendingMachineError.insufficientFunds(moneyNeeded: money)
        }
        // 구매하려는 수량보다 요구하는 수량이 많으면 오류를 던짐
        guard itemCount >= numberOfItemsToVend else {
            throw VendingMachineError.outOfStock
        }
        // 오류가 없으면 정상처리
        let totalPrice = numberOfItemsToVend * itemPrice

        self.deposited -= totalPrice
        self.itemCount -= numberOfItemsToVend

        return "\(numberOfItemsToVend)개 제공함"
    }
}

// 자판기 인스턴스
let machine: VendingMachine = VendingMachine()
// 판매 결과를 전달받을 변수
var result: String?
```

```swift
// 오류처리

// do-catch
do {
    try machine.receiveMoney(0)
}catch VendingMachineError.invalidInput {
    print("입력이 잘못되었습니다")
}catch VendingMachineError.insufficientFunds(let moneyNeeded) {
    print("\(moneyNeeded)원이 부족합니다")
}catch VendingMachineError.outOfStock {
    print("수량이 부족합니다")
} 입력이 잘못되었습니다

do {
    try machine.receiveMoney(300)
} catch {
// } catch (let error) { // 암시적 적용
    switch error {
        case VendingMachineError.invalidInput:
            print("입력이 잘못되었습니다")
                case VendingMachineError.insufficinetFunds(let moneyNeeded):
            print("\(moneyNeeded)원이 부족합니다")
                    case VendingMachineError.outOfStock:
            print("수량이 부족합니다")
        defalut:
            print("알 수 없는 오류 \(error)")
    }
} // 300원 받음

do {
    result = try machine.vend(numberOfItems: 4)
}catch {
    print(error)
} // insufficientFunds(100)

do {
    reslut = try machine.vend(numberOfItems: 4)
}
```

```swift
// try?
result = try? machine.vend(numberOfItems: 2)
result // Optional("2개 제공함")
result = try? machine.vend(numberOfItems: 2)
reslut // nil

//try!
// 옵셔널이 아닌 일반타입의 반환값을 내놓음
result = try! machine.vend(numberOfItems: 1)
result // 1개 제공함

// result = try! machine.vend(numberOfItems: 1)
```